src/html-imports.js

```diff
@@ -9,718 +9,754 @@
         try {
-          var locationHeader = request.getResponseHeader("Location");
-          if (locationHeader) {
-            redirectedUrl = (locationHeader.substr( 0, 1 ) === "/")
-              ? location.origin + locationHeader  // Location is a relative path
-              : locationHeader;                   // Full path
+          const u = new URL('b', 'http://a');
+          u.pathname = 'c%20d';
+          Path.__workingURL = (u.href === 'http://a/c%20d');
+        } catch (e) {}
+      }
+
+      if (Path.__workingURL) {
+        return (new URL(url, base)).href;
+      }
+
+      // Fallback to creating an anchor into a disconnected document.
+      let doc = Path.__tempDoc;
+      if (!doc) {
+        doc = document.implementation.createHTMLDocument('temp');
+        Path.__tempDoc = doc;
+        doc.__base = doc.createElement('base');
+        doc.head.appendChild(doc.__base);
+        doc.__anchor = doc.createElement('a');
+      }
+      doc.__base.href = base;
+      doc.__anchor.href = url;
+      return doc.__anchor.href || url;
+    }
+  };
+
+  /********************* Xhr processor *********************/
+  const Xhr = {
+
+    async: true,
+
+    /**
+     * @param {!string} url
+     * @param {!function(boolean, ?, string=)} callback
+     * @return {XMLHttpRequest}
+     */
+    load: function(url, callback) {
+      const request = new XMLHttpRequest();
+      if (flags.bust) {
+        url += '?' + Math.random();
+      }
+      request.open('GET', url, Xhr.async);
+      request.addEventListener('readystatechange', (e) => {
+        if (request.readyState === 4) {
+          // Servers redirecting an import can add a Location header to help us
+          // polyfill correctly.
+          let redirectedUrl = undefined;
+          try {
+            const locationHeader = request.getResponseHeader('Location');
+            if (locationHeader) {
+              // Relative or full path.
+              redirectedUrl = (locationHeader.substr(0, 1) === '/') ?
+                location.origin + locationHeader : locationHeader;
+            }
+          } catch (e) {
+            console.error(e.message);
           }
-        } catch ( e ) {
-            console.error( e.message );
+          const isOk = ((request.status >= 200 && request.status < 300) ||
+            request.status === 304 || request.status === 0);
+          const resource = (request.response || request.responseText);
+          callback(!isOk, resource, redirectedUrl);
         }
-        next.call(nextContext, !xhr.ok(request) && request,
-            request.response || request.responseText, redirectedUrl);
+      });
+      request.send();
+      return request;
+    }
+  };
+
+  /********************* loader *********************/
+  // This loader supports a dynamic list of urls
+  // and an oncomplete callback that is called when the loader is done.
+  // NOTE: The polyfill currently does *not* need this dynamism or the
```

Comment:
That's part of the potential improvement that could be done (in a separate PR). I've given it a shot to make it all promises-based but stumbled upon deduping issues, so I've postponed that work.